### DISCUSSION TITLE
How to add n different point sources?

### DISCUSSION THREAD

**Discussion by smallriver666** [SeisSol/SeisSol#835] [Q&A] [answered]
According to the referenced [WP2_LOH1](https://github.com/SeisSol/Examples/tree/master/WP2_LOH1) example, I can only add one point source. If I want to add several different point sources, what format should I follow? Is there any reference?
many thanks!

**Comment by daisy20170101** [comment]
there are several receivers in the file recordPoints. dat and should be written in the output. Did it not work like this in your case?

**Reply by smallriver666** [reply]
> there are several receivers in the file recordPoints. dat and should be written in the output. Did it not work like this in your case?
I once asked how to add `recordpoint.dat` files to the parameter file, but sebwolf-de told me that the [time-reversal implementation is experimental](https://github.com/SeisSol/SeisSol/discussions/831), so I don't know how to add the receiver.dat files.

**Comment by daisy20170101** [comment]
receivers are written in recordPoints.dat, which is called in parameters.par in this line:

RFileName = 'recordPoints.dat'       ! Record Points in extra file

**Comment by davschneller** [comment, marked as answer]
Hi,

you can do the following:

* for format 50 (Finite Source Rupture Model), see the script https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source_type50.py . To add more than one fault, you'll need to: exchange the 1 in the line after `number of faults` by the number of point sources. Then, add as many lines as you have faults after the `x,y,z,strike, dip, rake, area, onset time.` line (in the given script, only one line is generated, because there is only one fault). After ` source time function: dela, total sample point`, you'll stay at one line only (as the timestep and number of sample points is the same for all point sources in this format). After the line with `samples`, you'll need to add the said samples: first all samples for the first point source, then (without any empty line in between) the samples for the second point source and so on and so forth. Just as a hint, the format is rather sensitive to line breaks, so it's better to stick to the given format.
* for format 42, see https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source.py . Also, see https://seissol.readthedocs.io/en/latest/standard-rupture-format.html . The format description PDFs are seemingly unfortunately not available directly under the given links anymore, but the Internet Archive helps. See https://web.archive.org/web/20220122215640/http://hypocenter.usc.edu/research/SRF/srf4.pdf and https://web.archive.org/web/20220122215613/http://hypocenter.usc.edu/research/SRF/srf-v2.0_rev1.pdf . Both formats are supported by `rconv` (but note that the described `PLANE` block of the format is not supported). To add more points here, you can do the following: in line 33 of the script (I mean https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source.py by that), change `POINTS 1` to `POINTS N`, where N is the number of sources you have. Then for each of the N points, run through lines 34 to 36 with the adjusted values. That is, you have parameters, then samples for the first source; THEN the parameters for the second source, then the samples for the second source, then the parameters for the third source etc. For more infos, you can check out the source code of `rconv`, here: https://github.com/SeisSol/SeisSol/blob/master/preprocessing/science/rconv/src/SRF.cpp .

**Reply by smallriver666** [reply]
> Hi,
> 
> you can do the following:
> 
> * for format 50 (Finite Source Rupture Model), see the script https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source_type50.py . To add more than one fault, you'll need to: exchange the 1 in the line after `number of faults` by the number of point sources. Then, add as many lines as you have faults after the `x,y,z,strike, dip, rake, area, onset time.` line (in the given script, only one line is generated, because there is only one fault). After ` source time function: dela, total sample point`, you'll stay at one line only (as the timestep and number of sample points is the same for all point sources in this format). After the line with `samples`, you'll need to add the said samples: first all samples for the first point source, then (without any empty line in between) the samples for the second point source and so on and so forth. Just as a hint, the format is rather sensitive to line breaks, so it's better to stick to the given format.
> * for format 42, see https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source.py . Also, see https://seissol.readthedocs.io/en/latest/standard-rupture-format.html . The format description PDFs are seemingly unfortunately not available directly under the given links anymore, but the Internet Archive helps. See https://web.archive.org/web/20220122215640/http://hypocenter.usc.edu/research/SRF/srf4.pdf and https://web.archive.org/web/20220122215613/http://hypocenter.usc.edu/research/SRF/srf-v2.0_rev1.pdf . Both formats are supported by `rconv` (but note that the described `PLANE` block of the format is not supported). To add more points here, you can do the following: in line 33 of the script (I mean https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source.py by that), change `POINTS 1` to `POINTS N`, where N is the number of sources you have. Then for each of the N points, run through lines 34 to 36 with the adjusted values. That is, you have parameters, then samples for the first source; THEN the parameters for the second source, then the samples for the second source, then the parameters for the third source etc. For more infos, you can check out the source code of `rconv`, here: https://github.com/SeisSol/SeisSol/blob/master/preprocessing/science/rconv/src/SRF.cpp .

There is my [script](https://github.com/smallriver666/nPointSource/blob/main/generate_LOH_source_type50.py),  would you please tell me where is the problem with my script? 
I find the problem, I repeatedly added this [line](https://github.com/smallriver666/nPointSource/blob/main/generate_LOH_source_type50.py#:~:text=0.00000%20%201.0%20%20%20%20%200.00000-,x%2Cy%2Cz%2Cstrike%2C%20dip%2C%20rake%2C%20area%2C%20onset%20time.,-1000.00000%20%20%20%20%200.00000%20%20%20%20%200.00000).

**Reply by smallriver666** [reply]
> > Hi,
> > you can do the following:
> > 
> > * for format 50 (Finite Source Rupture Model), see the script https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source_type50.py . To add more than one fault, you'll need to: exchange the 1 in the line after `number of faults` by the number of point sources. Then, add as many lines as you have faults after the `x,y,z,strike, dip, rake, area, onset time.` line (in the given script, only one line is generated, because there is only one fault). After ` source time function: dela, total sample point`, you'll stay at one line only (as the timestep and number of sample points is the same for all point sources in this format). After the line with `samples`, you'll need to add the said samples: first all samples for the first point source, then (without any empty line in between) the samples for the second point source and so on and so forth. Just as a hint, the format is rather sensitive to line breaks, so it's better to stick to the given format.
> > * for format 42, see https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source.py . Also, see https://seissol.readthedocs.io/en/latest/standard-rupture-format.html . The format description PDFs are seemingly unfortunately not available directly under the given links anymore, but the Internet Archive helps. See https://web.archive.org/web/20220122215640/http://hypocenter.usc.edu/research/SRF/srf4.pdf and https://web.archive.org/web/20220122215613/http://hypocenter.usc.edu/research/SRF/srf-v2.0_rev1.pdf . Both formats are supported by `rconv` (but note that the described `PLANE` block of the format is not supported). To add more points here, you can do the following: in line 33 of the script (I mean https://github.com/SeisSol/Examples/blob/master/WP2_LOH1/generate_LOH_source.py by that), change `POINTS 1` to `POINTS N`, where N is the number of sources you have. Then for each of the N points, run through lines 34 to 36 with the adjusted values. That is, you have parameters, then samples for the first source; THEN the parameters for the second source, then the samples for the second source, then the parameters for the third source etc. For more infos, you can check out the source code of `rconv`, here: https://github.com/SeisSol/SeisSol/blob/master/preprocessing/science/rconv/src/SRF.cpp .
> 
> There is my [script](https://github.com/smallriver666/nPointSource/blob/main/generate_LOH_source_type50.py), would you please tell me where is the problem with my script? I find the problem, I repeatedly added this [line](https://github.com/smallriver666/nPointSource/blob/main/generate_LOH_source_type50.py#:~:text=0.00000%20%201.0%20%20%20%20%200.00000-,x%2Cy%2Cz%2Cstrike%2C%20dip%2C%20rake%2C%20area%2C%20onset%20time.,-1000.00000%20%20%20%20%200.00000%20%20%20%20%200.00000).

I still have a question, does this format of Point Source support adding in different media, such as adding one Point Source to water and one to rock, due to the `seismic moment tensor` of the acoustic wave equation in water and the elastic wave equation in rock It is different, so if I want to add a different seismic moment tensor, what should I do, Thanks!
